File: android/src/main/java/com/bundledrop/BundleDropModule.kt
package com.bundledrop

import android.content.Context
import android.content.Intent
import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.modules.network.OkHttpClientProvider
import java.io.File
import org.json.JSONArray
import org.json.JSONObject

class BundleDropModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  companion object {
    fun getDownloadedBundlePath(context: Context): String? =
      BundleDropNativePaths.getDownloadedBundlePathPassive(context)

    @JvmStatic
    fun resolveJSBundleFile(context: Context, fallback: String?): String? {
      val path = BundleDropNativePaths.getDownloadedBundlePath(context)
      return if (!path.isNullOrEmpty()) path else fallback
    }
  }

  override fun getName(): String = "BundleDrop"

  @ReactMethod
  fun getDownloadedBundlePath(promise: Promise) {
    promise.resolve(getDownloadedBundlePath(reactApplicationContext))
  }

  @ReactMethod
  fun restartReactNative() {
    val activity = reactApplicationContext.currentActivity ?: return
    val context = reactApplicationContext

    val intent = context.packageManager.getLaunchIntentForPackage(context.packageName)
    if (intent == null) {
      Log.e("BundleDrop", "❌ Could not get launch intent for package")
      return
    }

    try {
      intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
      activity.finish()
      context.startActivity(intent)
      Runtime.getRuntime().exit(0)

      Log.d("BundleDrop", "🔁 Restarting app with new bundle...")
    } catch (e: Exception) {
      Log.e("BundleDrop", "❌ Failed to restart app: ${e.message}", e)
    }
  }

  @ReactMethod
  fun setOtaEnabled(enabled: Boolean, promise: Promise) {
    try {
      BundleDropOtaPrefs.writeOtaEnabled(reactApplicationContext, enabled)
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("ERR_SET_OTA_ENABLED", e.message, e)
    }
  }

  @ReactMethod
  fun activateStartupCandidate(
    hash: String,
    maxCrashCount: Double,
    healthCheckMode: String,
    healthyAfterSec: Double,
    promise: Promise,
  ) {
    try {
      require(maxCrashCount.isFinite() && maxCrashCount >= 0 && maxCrashCount % 1.0 == 0.0) {
        "maxCrashCount must be a non-negative integer"
      }
      require(maxCrashCount <= Int.MAX_VALUE) { "maxCrashCount is outside the supported range" }
      val result = BundleDropStartupRecovery.controller(reactApplicationContext)
        .activateCandidate(hash, maxCrashCount.toInt(), healthCheckMode, healthyAfterSec)
      promise.resolve(Arguments.createMap().apply {
        putString("hash", result.hash)
        putString("bundlePath", result.bundlePath)
      })
    } catch (error: Exception) {
      promise.reject("ERR_STARTUP_RECOVERY_ACTIVATE", error.message, error)
    }
  }

  @ReactMethod
  fun markStartupHealthy(hash: String, attemptId: String, promise: Promise) {
    try {
      val marked = BundleDropStartupRecovery.controller(reactApplicationContext)
        .markHealthy(hash, attemptId)
      promise.resolve(marked)
    } catch (error: Exception) {
      promise.reject("ERR_STARTUP_RECOVERY_HEALTH", error.message, error)
    }
  }

  @ReactMethod
  fun getStartupRecoveryState(promise: Promise) {
    try {
      val state = BundleDropStartupRecovery.controller(reactApplicationContext).snapshot()
      promise.resolve(jsonObjectToWritableMap(state))
    } catch (error: Exception) {
      promise.reject("ERR_STARTUP_RECOVERY_STATE", error.message, error)
    }
  }

  @ReactMethod
  fun setStartupRecoveryRevokedHashes(hashes: ReadableArray, promise: Promise) {
    try {
      val values = buildSet {
        for (index in 0 until hashes.size()) {
          val hash = hashes.getString(index)
            ?: throw IllegalArgumentException("Revoked bundle hashes must be strings")
          add(hash)
        }
      }
      promise.resolve(
        BundleDropStartupRecovery.controller(reactApplicationContext)
          .setRevokedHashes(values),
      )
    } catch (error: Exception) {
      promise.reject("ERR_STARTUP_RECOVERY_REVOKE", error.message, error)
    }
  }

  @ReactMethod
  fun acknowledgeStartupRecovery(eventId: String, promise: Promise) {
    try {
      promise.resolve(
        BundleDropStartupRecovery.controller(reactApplicationContext)
          .acknowledgeRecovery(eventId),
      )
    } catch (error: Exception) {
      promise.reject("ERR_STARTUP_RECOVERY_ACK", error.message, error)
    }
  }

  @ReactMethod
  fun rollbackStartupBundle(forceEmbedded: Boolean, promise: Promise) {
    try {
      val result = BundleDropStartupRecovery.controller(reactApplicationContext)
        .rollbackStartupBundle(forceEmbedded)
      promise.resolve(Arguments.createMap().apply {
        putBoolean("rolledBack", result.rolledBack)
        putBoolean("toEmbedded", result.toEmbedded)
        result.hash?.let { putString("hash", it) }
      })
    } catch (error: Exception) {
      promise.reject("ERR_STARTUP_RECOVERY_ROLLBACK", error.message, error)
    }
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun getImageManifestSync(): String? {
    return readImageManifestRaw()
  }

  @ReactMethod
  fun getImageManifest(promise: Promise) {
    promise.resolve(readImageManifestRaw())
  }

  private fun readImageManifestRaw(): String? {
    val context = reactApplicationContext
    val bundleDropRoot = File(context.filesDir, "bundle-drop")
    val current = BundleDropOtaResolver.readCurrentPointer(bundleDropRoot) ?: return null
    return BundleDropOtaResolver.readImageManifest(current)
  }

  // ---------------------------------------------------------------------------
  // FS primitives (replaces react-native-fs dependency)
  // ---------------------------------------------------------------------------

  @ReactMethod
  fun fsExists(path: String, promise: Promise) {
    promise.resolve(File(path).exists())
  }

  @ReactMethod
  fun fsReadFile(path: String, encoding: String, promise: Promise) {
    try {
      val file = File(path)
      if (!file.exists()) {
        promise.reject("ENOENT", "File not found: $path")
        return
      }
      if (encoding == "base64") {
        val bytes = file.readBytes()
        promise.resolve(android.util.Base64.encodeToString(bytes, android.util.Base64.NO_WRAP))
      } else {
        promise.resolve(file.readText(Charsets.UTF_8))
      }
    } catch (e: Exception) {
      promise.reject("ERR_READ", e.message, e)
    }
  }

  @ReactMethod
  fun fsWriteFile(path: String, content: String, encoding: String, promise: Promise) {
    try {
      val file = File(path)
      file.parentFile?.mkdirs()
      if (encoding == "base64") {
        val bytes = android.util.Base64.decode(content, android.util.Base64.DEFAULT)
        file.writeBytes(bytes)
      } else {
        file.writeText(content, Charsets.UTF_8)
      }
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("ERR_WRITE", e.message, e)
    }
  }

  @ReactMethod
  fun fsMkdir(path: String, promise: Promise) {
    try {
      val dir = File(path)
      if (!dir.exists()) dir.mkdirs()
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("ERR_MKDIR", e.message, e)
    }
  }

  @ReactMethod
  fun fsReadDir(path: String, promise: Promise) {
    try {
      val dir = File(path)
      val names = dir.list() ?: emptyArray()
      val result = Arguments.createArray()
      names.forEach { result.pushString(it) }
      promise.resolve(result)
    } catch (e: Exception) {
      promise.reject("ERR_READDIR", e.message, e)
    }
  }

  @ReactMethod
  fun fsUnlink(path: String, promise: Promise) {
    try {
      BundleDropFileOps.unlinkPath(path)
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("ERR_UNLINK", e.message, e)
    }
  }

  @ReactMethod
  fun fsMoveFile(src: String, dest: String, promise: Promise) {
    try {
      BundleDropFileOps.moveFile(File(src), File(dest))
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("ERR_MOVE", e.message, e)
    }
  }

  @ReactMethod
  fun fsCopyFile(src: String, dest: String, promise: Promise) {
    try {
      BundleDropFileOps.copyFile(File(src), File(dest))
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("ERR_COPY", e.message, e)
    }
  }

  @ReactMethod
  fun fsSha256File(path: String, promise: Promise) {
    try {
      promise.resolve(BundleDropFileOps.sha256File(File(path)))
    } catch (e: Exception) {
      promise.reject("ERR_SHA256", e.message, e)
    }
  }

  @ReactMethod
  fun fsSha256String(value: String, promise: Promise) {
    try {
      promise.resolve(BundleDropRuntimeCrypto.sha256String(value))
    } catch (e: Exception) {
      promise.reject("ERR_SHA256", e.message, e)
    }
  }

  @ReactMethod
  fun fsVerifyEs256Signature(
    signingInput: String,
    signatureBase64Url: String,
    xBase64Url: String,
    yBase64Url: String,
    promise: Promise,
  ) {
    try {
      promise.resolve(
        BundleDropRuntimeCrypto.verifyEs256Signature(
          signingInput,
          signatureBase64Url,
          xBase64Url,
          yBase64Url,
        ),
      )
    } catch (e: Exception) {
      promise.reject("ERR_ES256_VERIFY", e.message, e)
    }
  }

  @ReactMethod
  fun fsFileSize(path: String, promise: Promise) {
    try {
      promise.resolve(BundleDropFileOps.fileSize(File(path)).toDouble())
    } catch (e: Exception) {
      promise.reject("ERR_FILE_SIZE", e.message, e)
    }
  }

  @ReactMethod
  fun fsApplyXdelta(basePath: String, patchPath: String, outputPath: String, promise: Promise) {
    Thread {
      try {
        BundleDropFileOps.applyXdelta(File(basePath), File(patchPath), File(outputPath))
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("ERR_XDELTA", e.message, e)
      }
    }.start()
  }

  @ReactMethod
  fun fsVerifyBundleFiles(bundleDir: String, manifestPath: String, promise: Promise) {
    Thread {
      try {
        BundleDropBundleVerifier.verifyBundleFiles(File(bundleDir), File(manifestPath))
        val result = Arguments.createMap()
        result.putBoolean("verified", true)
        promise.resolve(result)
      } catch (e: Exception) {
        promise.reject("ERR_VERIFY_BUNDLE", e.message, e)
      }
    }.start()
  }

  @ReactMethod
  fun fsSupportsXdelta(promise: Promise) {
    promise.resolve(BundleDropFileOps.supportsXdelta())
  }

  @ReactMethod
  fun fsUnzip(zipPath: String, destPath: String, promise: Promise) {
    Thread {
      try {
        val filenames = BundleDropFileOps.unzipFile(zipPath, destPath)
        val result = Arguments.createArray()
        filenames.forEach { result.pushString(it) }
        promise.resolve(result)
      } catch (e: Exception) {
        promise.reject("ERR_UNZIP", e.message, e)
      }
    }.start()
  }

  @ReactMethod
  fun fsDownloadFile(url: String, destPath: String, promise: Promise) {
    Thread {
      try {
        val parsedUrl = BundleDropFileOps.validateHttpUrl(url)
        BundleDropFileOps.downloadToFile(parsedUrl, File(destPath))
        promise.resolve(null)
      } catch (e: Exception) {
        try { File(destPath).delete() } catch (_: Exception) {}
        promise.reject("ERR_DOWNLOAD", e.message, e)
      }
    }.start()
  }

  @ReactMethod
  fun fsDownloadFileBounded(
    url: String,
    destPath: String,
    maxBytes: Double,
    timeoutMs: Double,
    promise: Promise,
  ) {
    Thread {
      try {
        require(maxBytes.isFinite() && maxBytes >= 1) { "maxBytes must be positive" }
        require(timeoutMs.isFinite() && timeoutMs >= 1 && timeoutMs <= Int.MAX_VALUE) {
          "timeoutMs is outside the supported range"
        }
        val parsedUrl = BundleDropFileOps.validateHttpUrl(url)
        BundleDropBoundedDownloader.downloadToFile(
          OkHttpClientProvider.getOkHttpClient(),
          parsedUrl,
          File(destPath),
          maxBytes = maxBytes.toLong(),
          timeoutMs = timeoutMs.toLong(),
        )
        promise.resolve(null)
      } catch (e: Exception) {
        try { File(destPath).delete() } catch (_: Exception) {}
        val code = when {
          e is BundleDropBoundedDownloadTimeoutException -> "ERR_DOWNLOAD_TIMEOUT"
          e is BundleDropBoundedDownloadTooLargeException -> "ERR_DOWNLOAD_TOO_LARGE"
          e is BundleDropBoundedDownloadHttpException -> "ERR_DOWNLOAD_HTTP"
          else -> "ERR_DOWNLOAD_NETWORK"
        }
        val message = e.message.orEmpty()
        promise.reject(code, message.ifEmpty { "Manifest download failed" }, e)
      }
    }.start()
  }

  override fun getConstants(): MutableMap<String, Any> {
    val map = mutableMapOf<String, Any>()
    map["startupRecoveryProtocolVersion"] = BundleDropStartupRecoveryController.PROTOCOL_VERSION
    @Suppress("UNCHECKED_CAST")
    (map as MutableMap<String, Any?>)["startupRecoverySelectedHash"] =
      BundleDropStartupRecovery.startupSelectedHash()
    BundleDropStartupRecovery.startupAttempt()?.let { (hash, attemptId) ->
      map["startupRecoveryAttemptHash"] = hash
      map["startupRecoveryAttemptId"] = attemptId
    }
    val context = reactApplicationContext
    map["DocumentDirectoryPath"] = context.filesDir.absolutePath
    map["LibraryDirectoryPath"] = context.filesDir.absolutePath
    return map
  }

  private fun jsonObjectToWritableMap(json: JSONObject): WritableMap {
    val map = Arguments.createMap()
    val keys = json.keys()
    while (keys.hasNext()) {
      val key = keys.next()
      putJsonValue(map, key, json.opt(key))
    }
    return map
  }

  private fun jsonArrayToWritableArray(json: JSONArray): WritableArray {
    val array = Arguments.createArray()
    for (index in 0 until json.length()) {
      when (val value = json.opt(index)) {
        null, JSONObject.NULL -> array.pushNull()
        is JSONObject -> array.pushMap(jsonObjectToWritableMap(value))
        is JSONArray -> array.pushArray(jsonArrayToWritableArray(value))
        is Boolean -> array.pushBoolean(value)
        is Number -> array.pushDouble(value.toDouble())
        else -> array.pushString(value.toString())
      }
    }
    return array
  }

  private fun putJsonValue(map: WritableMap, key: String, value: Any?) {
    when (value) {
      null, JSONObject.NULL -> map.putNull(key)
      is JSONObject -> map.putMap(key, jsonObjectToWritableMap(value))
      is JSONArray -> map.putArray(key, jsonArrayToWritableArray(value))
      is Boolean -> map.putBoolean(key, value)
      is Number -> map.putDouble(key, value.toDouble())
      else -> map.putString(key, value.toString())
    }
  }
}

File: ios/BundleDropModule.swift
import Foundation
import React

@objc(BundleDropModule)
final class BundleDropModule: NSObject {

  private static var isReloading = false
  private static let fm = FileManager.default
  private static let fileOpsQueue = DispatchQueue(label: "com.bundledrop.file-ops", qos: .userInitiated)

  private func runFileOperation(
    errorCode: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock,
    operation: @escaping () throws -> Any?
  ) {
    Self.fileOpsQueue.async {
      do {
        resolve(try operation())
      } catch {
        reject(errorCode, error.localizedDescription, error)
      }
    }
  }

  // MARK: - OTA bundle path

  @objc
  func getDownloadedBundlePath(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    if let url = BundleDropStartupRecoveryAdapter.downloadedBundleURL() {
      print("📦 Found downloaded bundle at: \(url.path)")
      resolve(url.path)
      return
    }
    print("📦 No downloaded bundle found.")
    resolve(nil)
  }

  @objc
  func activateStartupCandidate(
    _ hash: String,
    maxCrashCount: NSNumber,
    healthCheckMode: String,
    healthyAfterSec: NSNumber,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_STARTUP_RECOVERY_ACTIVATE", resolve: resolve, reject: reject) {
      let crashCount = maxCrashCount.doubleValue
      let healthDelay = healthyAfterSec.doubleValue
      guard crashCount.isFinite,
            crashCount >= 0,
            crashCount.rounded(.towardZero) == crashCount,
            crashCount <= Double(Int32.max),
            healthDelay.isFinite,
            healthDelay >= 0 else {
        throw BundleDropStartupRecoveryError.invalidPolicy
      }
      let result = try BundleDropStartupRecoveryAdapter.activateCandidate(
        hash: hash,
        maxCrashCount: Int(crashCount),
        healthCheckMode: healthCheckMode,
        healthyAfterSec: healthDelay
      )
      return ["hash": result.hash, "bundlePath": result.bundleURL.path]
    }
  }

  @objc
  func markStartupHealthy(
    _ hash: String,
    attemptId: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_STARTUP_RECOVERY_HEALTH", resolve: resolve, reject: reject) {
      BundleDropStartupRecoveryAdapter.markHealthy(hash: hash, attemptId: attemptId)
    }
  }

  @objc
  func getStartupRecoveryState(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_STARTUP_RECOVERY_STATE", resolve: resolve, reject: reject) {
      try BundleDropStartupRecoveryAdapter.snapshot()
    }
  }

  @objc
  func setStartupRecoveryRevokedHashes(
    _ hashes: [String],
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_STARTUP_RECOVERY_REVOKE", resolve: resolve, reject: reject) {
      try BundleDropStartupRecoveryAdapter.setRevokedHashes(hashes)
    }
  }

  @objc
  func acknowledgeStartupRecovery(
    _ eventId: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_STARTUP_RECOVERY_ACK", resolve: resolve, reject: reject) {
      try BundleDropStartupRecoveryAdapter.acknowledgeRecovery(eventId: eventId)
    }
  }

  @objc
  func rollbackStartupBundle(
    _ forceEmbedded: Bool,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_STARTUP_RECOVERY_ROLLBACK", resolve: resolve, reject: reject) {
      let result = try BundleDropStartupRecoveryAdapter.rollback(forceEmbedded: forceEmbedded)
      var response: [String: Any] = [
        "rolledBack": result.rolledBack,
        "toEmbedded": result.toEmbedded,
      ]
      if let hash = result.hash {
        response["hash"] = hash
      }
      return response
    }
  }

  @objc
  func setOtaEnabled(
    _ enabled: Bool,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    UserDefaults.standard.set(enabled, forKey: BundleDropLocatorCore.otaEnabledKey)
    resolve(nil)
  }

  @objc
  func restartReactNative() {
    DispatchQueue.main.async {
      if Self.isReloading {
        print("🔁 Restart already in progress - ignoring.")
        return
      }

      guard let stagedURL = BundleDropStartupRecoveryAdapter.downloadedBundleURL() else {
        print("⚠️ No downloaded bundle found; skipping restart.")
        return
      }

      let size = BundleDropLocatorCore.fileSize(at: stagedURL)
      if size < 1024 {
        print("⚠️ Bundle exists but looks invalid (size=\(size)). Skipping restart.")
        return
      }

      guard let url = BundleDropStartupRecoveryAdapter.beginReload() else {
        print("⚠️ Downloaded bundle became unavailable; skipping restart.")
        return
      }

      Self.isReloading = true
      print("🔄 Restarting RN from \(url.path) size=\(size)")

      RCTReloadCommandSetBundleURL(url)
      RCTTriggerReloadCommandListeners("bundle-drop-restart")

      DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
        Self.isReloading = false
      }
    }
  }

  // MARK: - Constants

  @objc
  func constantsToExport() -> [String: Any] {
    let fm = FileManager.default
    let docs = fm.urls(for: .documentDirectory, in: .userDomainMask).first?.path ?? ""
    let lib = fm.urls(for: .libraryDirectory, in: .userDomainMask).first?.path ?? ""
    let attempt = BundleDropStartupRecoveryAdapter.capturedAttempt()
    return [
      "DocumentDirectoryPath": docs,
      "LibraryDirectoryPath": lib,
      "startupRecoveryProtocolVersion": BundleDropStartupRecoveryAdapter.protocolVersion,
      "startupRecoverySelectedHash": BundleDropStartupRecoveryAdapter.capturedSelectedHash() ?? NSNull(),
      "startupRecoveryAttemptHash": attempt.hash ?? NSNull(),
      "startupRecoveryAttemptId": attempt.attemptId ?? NSNull(),
    ]
  }

  // MARK: - FS primitives (replaces react-native-fs dependency)

  @objc
  func fsExists(
    _ path: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(Self.fm.fileExists(atPath: path))
  }

  @objc
  func fsReadFile(
    _ path: String,
    encoding: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    guard Self.fm.fileExists(atPath: path) else {
      reject("ENOENT", "File not found: \(path)", nil)
      return
    }
    do {
      let fileURL = URL(fileURLWithPath: path)
      if encoding == "base64" {
        let data = try Data(contentsOf: fileURL)
        resolve(data.base64EncodedString())
      } else {
        let text = try String(contentsOf: fileURL, encoding: .utf8)
        resolve(text)
      }
    } catch {
      reject("ERR_READ", error.localizedDescription, error)
    }
  }

  @objc
  func fsWriteFile(
    _ path: String,
    content: String,
    encoding: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      let dir = (path as NSString).deletingLastPathComponent
      if !Self.fm.fileExists(atPath: dir) {
        try Self.fm.createDirectory(atPath: dir, withIntermediateDirectories: true)
      }
      let fileURL = URL(fileURLWithPath: path)
      if encoding == "base64" {
        guard let data = Data(base64Encoded: content) else {
          reject("ERR_WRITE", "Invalid base64 data", nil)
          return
        }
        try data.write(to: fileURL)
      } else {
        try content.write(to: fileURL, atomically: true, encoding: .utf8)
      }
      resolve(nil)
    } catch {
      reject("ERR_WRITE", error.localizedDescription, error)
    }
  }

  @objc
  func fsMkdir(
    _ path: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      if !Self.fm.fileExists(atPath: path) {
        try Self.fm.createDirectory(atPath: path, withIntermediateDirectories: true)
      }
      resolve(nil)
    } catch {
      reject("ERR_MKDIR", error.localizedDescription, error)
    }
  }

  @objc
  func fsUnlink(
    _ path: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      try BundleDropFileOps.unlinkPath(path)
      resolve(nil)
    } catch {
      reject("ERR_UNLINK", error.localizedDescription, error)
    }
  }

  @objc
  func fsReadDir(
    _ path: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      let items = try Self.fm.contentsOfDirectory(atPath: path)
      resolve(items)
    } catch {
      reject("ERR_READDIR", error.localizedDescription, error)
    }
  }

  @objc
  func fsMoveFile(
    _ src: String,
    dest: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_MOVE", resolve: resolve, reject: reject) {
      try BundleDropFileOps.moveFile(srcPath: src, destPath: dest)
      return nil
    }
  }

  @objc
  func fsCopyFile(
    _ src: String,
    dest: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_COPY", resolve: resolve, reject: reject) {
      try BundleDropFileOps.copyFile(srcPath: src, destPath: dest)
      return nil
    }
  }

  @objc
  func fsSha256File(
    _ path: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_SHA256", resolve: resolve, reject: reject) {
      try BundleDropFileOps.sha256File(path)
    }
  }

  @objc
  func fsSha256String(
    _ value: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(BundleDropRuntimeCrypto.sha256String(value))
  }

  @objc
  func fsVerifyEs256Signature(
    _ signingInput: String,
    signatureBase64Url: String,
    xBase64Url: String,
    yBase64Url: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      resolve(try BundleDropRuntimeCrypto.verifyEs256Signature(
        signingInput: signingInput,
        signatureBase64Url: signatureBase64Url,
        xBase64Url: xBase64Url,
        yBase64Url: yBase64Url
      ))
    } catch {
      reject("ERR_ES256_VERIFY", error.localizedDescription, error)
    }
  }

  @objc
  func fsFileSize(
    _ path: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      resolve(try BundleDropFileOps.fileSize(path))
    } catch {
      reject("ERR_FILE_SIZE", error.localizedDescription, error)
    }
  }

  @objc
  func fsApplyXdelta(
    _ basePath: String,
    patchPath: String,
    outputPath: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_XDELTA", resolve: resolve, reject: reject) {
      try BundleDropFileOps.applyXdelta(
        basePath: basePath,
        patchPath: patchPath,
        outputPath: outputPath
      )
      return nil
    }
  }

  @objc
  func fsVerifyBundleFiles(
    _ bundleDir: String,
    manifestPath: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_VERIFY_BUNDLE", resolve: resolve, reject: reject) {
      try BundleDropBundleVerifier.verifyBundleFiles(
        bundleDir: URL(fileURLWithPath: bundleDir),
        manifestURL: URL(fileURLWithPath: manifestPath)
      )
      return ["verified": true]
    }
  }

  @objc
  func fsSupportsXdelta(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(BundleDropFileOps.supportsXdelta())
  }

  @objc
  func fsUnzip(
    _ zipPath: String,
    destPath: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    runFileOperation(errorCode: "ERR_UNZIP", resolve: resolve, reject: reject) {
      let filenames = try BundleDropZipExtractor.extractZip(atPath: zipPath,
                                                            toDirectory: destPath)
      return filenames ?? []
    }
  }

  @objc
  func fsDownloadFile(
    _ url: String,
    destPath: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    let config = URLSessionConfiguration.default
    config.timeoutIntervalForRequest = 30
    config.timeoutIntervalForResource = 300

    BundleDropFileOps.downloadToFile(
      urlString: url,
      destPath: destPath,
      configuration: config
    ) { result in
      switch result {
      case .success:
        resolve(nil)
      case .failure(let error):
        reject("ERR_DOWNLOAD", error.localizedDescription, error)
      }
    }
  }

  @objc
  func fsDownloadFileBounded(
    _ url: String,
    destPath: String,
    maxBytes: NSNumber,
    timeoutMs: NSNumber,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    let maxByteCount = maxBytes.int64Value
    let timeoutSeconds = timeoutMs.doubleValue / 1000
    guard maxByteCount > 0, timeoutSeconds > 0 else {
      reject("ERR_DOWNLOAD_NETWORK", "Invalid bounded download limits", nil)
      return
    }

    let config = URLSessionConfiguration.default
    config.timeoutIntervalForRequest = timeoutSeconds
    config.timeoutIntervalForResource = timeoutSeconds
    BundleDropFileOps.downloadToFile(
      urlString: url,
      destPath: destPath,
      maxBytes: maxByteCount,
      configuration: config
    ) { result in
      switch result {
      case .success:
        resolve(nil)
      case .failure(let error):
        let message = error.localizedDescription
        let errorCode: String
        if (error as? URLError)?.code == .timedOut {
          errorCode = "ERR_DOWNLOAD_TIMEOUT"
        } else if message.hasPrefix("Download exceeds ") {
          errorCode = "ERR_DOWNLOAD_TOO_LARGE"
        } else if message.range(of: #"^HTTP \d{3}(?:\b|:)"#, options: .regularExpression) != nil {
          errorCode = "ERR_DOWNLOAD_HTTP"
        } else {
          errorCode = "ERR_DOWNLOAD_NETWORK"
        }
        reject(errorCode, message, error)
      }
    }
  }
}
